File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/listview/sidebar/FilterPanelController.java
package com.goexp.galgame.gui.view.game.listview.sidebar;

import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.PanelTask;
import com.goexp.galgame.gui.util.TaskService;
import com.goexp.galgame.gui.view.game.listview.sidebar.node.BrandItemNode;
import com.goexp.galgame.gui.view.game.listview.sidebar.node.CompItemNode;
import com.goexp.galgame.gui.view.game.listview.sidebar.node.DateItemNode;
import com.goexp.galgame.gui.view.game.listview.sidebar.node.DefaultItemNode;
import javafx.concurrent.Service;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.CheckBox;
import javafx.scene.control.TreeCell;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeView;
import javafx.scene.layout.FlowPane;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class FilterPanelController extends FilterController<Game> {

    @FXML
    private FlowPane starbox;

    @FXML
    private FlowPane statebox;

    protected void initialize() {

        reset();
    }

    public void reset() {
        var nodes = Stream.of(GameState.values())
                .sorted(Comparator.comparing(GameState::getValue).reversed())
                .map(gameType -> {
                    var btn = new CheckBox(gameType.getName());
                    btn.setUserData(gameType);

                    if (gameType.getValue() > GameState.BLOCK.getValue())
                        btn.setSelected(true);

                    return btn;
                }).collect(Collectors.toList());

        statebox.getChildren().setAll(nodes);


        var starnodes = IntStream.rangeClosed(0, 5)
                .boxed()
                .sorted(Comparator.comparing(Integer::intValue).reversed())
                .map(star -> {
                    var btn = new CheckBox(star.toString());

                    btn.setSelected(true);

                    btn.setUserData(star);

                    return btn;
                }).collect(Collectors.toList());

        starbox.getChildren().setAll(starnodes);

        setP();
    }

    @FXML
    private void SetFilter_OnAction(ActionEvent event) {
        setP();

        onSetProperty.set(true);
        onSetProperty.set(false);
    }

    private void setP() {
        var stars = starbox.getChildren().stream()
                .filter(node -> ((CheckBox) node).isSelected())
                .map(node -> Integer.parseInt(((CheckBox) node).getText()))
                .collect(Collectors.toSet());

        var states = statebox.getChildren().stream()
                .filter(node -> ((CheckBox) node).isSelected())
                .map(node -> (GameState) (node.getUserData()))
                .collect(Collectors.toSet());

        if (stars.isEmpty())
            stars.add(0);

        predicate = game -> stars.contains(game.star) && states.contains(game.state.get());
    }

    @Override
    public void init(List<Game> filteredGames) {

    }

    public static class BrandGroupController extends FilterController<Game> {

        @FXML
        private TreeView<DefaultItemNode> compTree;

        private List<Game> filteredGames;

        private Service<TreeItem<DefaultItemNode>> groupBrandServ = new TaskService<>(() -> new PanelTask.GroupBrand(filteredGames));


        protected void initialize() {
            compTree.setCellFactory(itemNodeTreeView -> {
                return new TreeCell<>() {
                    @Override
                    protected void updateItem(DefaultItemNode item, boolean empty) {
                        super.updateItem(item, empty);

                        setGraphic(null);
                        setText(null);

                        if (!empty && item != null) {

                            setText(String.format("%s (%d)", item.title, item.count));
                        }
                    }
                };
            });
            compTree.getSelectionModel().selectedItemProperty().addListener((o, old, item) -> {
                if (item != null) {

                    if (item.getValue() instanceof CompItemNode) {
                        predicate = game -> {
                            // target comp can be null,so...change null to empty
                            var comp = game.brand.comp != null ? game.brand.comp : "";

                            return comp.equals(((CompItemNode) item.getValue()).comp);
                        };

                    } else {
                        predicate = game -> (game.brand.equals(((BrandItemNode) item.getValue()).brand));
                    }

                    onSetProperty.set(true);
                    onSetProperty.set(false);
                }
            });

            groupBrandServ.valueProperty().addListener((observable, oldValue, newValue) -> {
                if (newValue != null) {
                    compTree.setRoot(newValue);
                }
            });

        }

        @Override
        public void init(List<Game> filteredGames) {
            this.filteredGames = filteredGames;

            groupBrandServ.restart();

        }


    }

    public static class DateGroupController extends FilterController<Game> {

        @FXML
        private TreeView<DateItemNode> dateTree;

        private List<Game> filteredGames;

        private Service<TreeItem<DateItemNode>> groupDateServ = new TaskService<>(() -> new PanelTask.GroupDate(filteredGames));


        protected void initialize() {

            dateTree.setCellFactory(dateItemNodeTreeView -> {
                return new TreeCell<>() {

                    @Override
                    protected void updateItem(DateItemNode item, boolean empty) {
                        super.updateItem(item, empty);

                        setGraphic(null);
                        setText(null);

                        if (!empty && item != null) {
                            setText(item.title);
                        }
                    }
                };
            });
            dateTree.getSelectionModel().selectedItemProperty().addListener((observable, oldValue, newValue) -> {

                if (newValue != null) {
                    predicate = game -> game.publishDate != null && game.publishDate.isBefore(newValue.getValue().range.end) && game.publishDate.isAfter(newValue.getValue().range.start);

                    onSetProperty.set(true);
                    onSetProperty.set(false);
                }
            });

            groupDateServ.valueProperty().addListener((observable, oldValue, newValue) -> {
                if (newValue != null) {
                    dateTree.setRoot(newValue);
                }
            });
        }

        public void init(List<Game> filteredGames) {
            this.filteredGames = filteredGames;

            groupDateServ.restart();
        }
    }


}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/listview/sidebar/FilterPanelController.scala
package com.goexp.galgame.gui.view.game.listview.sidebar

import java.time.LocalDate
import java.util

import com.goexp.galgame.common.model.game.{GameLocation, GameState}
import com.goexp.galgame.gui.model.Game
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control.CheckBox
import javafx.scene.layout.Pane
import org.controlsfx.control.ToggleSwitch

import scala.jdk.CollectionConverters._

class FilterPanelController extends FilterController[Game] {
  @FXML private var starbox: Pane = _
  @FXML private var statebox: Pane = _
  @FXML private var locationbox: Pane = _

  @FXML private var switchAll: ToggleSwitch = _

  override protected def initialize() = reset()

  def reset() = {

    def resetState() = {
      val nodes = GameState.values.to(LazyList)
        .sortBy {
          _.value
        }.reverse
        .map(gameType => {
          val btn = new CheckBox(gameType.name)
          btn.setUserData(gameType)
          if (gameType.value > GameState.BLOCK.value)
            btn.setSelected(true)
          btn
        })
        .asJava

      statebox.getChildren.setAll(nodes)
    }

    resetState()

    def resetStar() = {
      val nodes =

        Range.inclusive(0, 5).to(LazyList).reverse
          .map(star => {
            val btn = new CheckBox(star.toString)
            btn.setUserData(star)
            btn.setSelected(true)
            btn
          }).asJava


      starbox.getChildren.setAll(nodes)
    }

    resetStar()

    def resetLocation() = {
      val nodes = List(GameLocation.REMOTE, GameLocation.NETDISK, GameLocation.LOCAL).to(LazyList)
        .map(gameLoc => {
          val btn = new CheckBox(gameLoc.name)
          btn.setUserData(gameLoc)
          btn.setSelected(true)
          btn
        })
        .asJava

      locationbox.getChildren.setAll(nodes)
    }

    resetLocation()


    setP()
  }

  @FXML private def SetFilter_OnAction(event: ActionEvent) = {
    setP()
    onSetProperty.set(true)
    onSetProperty.set(false)
  }

  private def setP() = {
    val stars = starbox.getChildren.asScala.to(LazyList)
      .filter(_.asInstanceOf[CheckBox].isSelected)
      .map(_.asInstanceOf[CheckBox].getText.toInt)

    val states = statebox.getChildren.asScala.to(LazyList)
      .filter(_.asInstanceOf[CheckBox].isSelected)
      .map(_.getUserData.asInstanceOf[GameState])

    val location = locationbox.getChildren.asScala.to(LazyList)
      .filter(_.asInstanceOf[CheckBox].isSelected)
      .map(_.getUserData.asInstanceOf[GameLocation])

    predicate = (game: Game) => {
      stars.contains(game.star.get()) &&
        states.contains(game.state.get) &&
        location.contains(game.location.get)
    }

    if (switchAll.isSelected)
      predicate = predicate.and { game: Game =>
        Option(game.publishDate).exists {
          _.isBefore(LocalDate.now())
        }
      }
  }

  override def init(filteredGames: util.List[Game]) = {
  }
}